### PR TITLE
BLO-553 deploy smart contract

### PR DESCRIPTION
**Pull request by bluecco**

### CODE REVIEW

File: packages/extension/src/background/transactions/transactionMessaging.ts [hunk 1]
```diff
@@ -163,6 +163,48 @@ export const handleTransactionMessage: HandleMessage<
...
+          },
+        })
+      } catch (error) {
+        console.log(error)
+        return respond({
+          type: "ESTIMATE_DEPLOY_CONTRACT_FEE_RES",
+          data: {
+            error:
+              (error as any)?.message?.toString?.() ??
+              (error as any)?.toString?.() ??
```
Comment (delaaxe) [inline packages/extension/src/background/transactions/transactionMessaging.ts:201]: `.toString()` is always defined on non-nullable objects

```suggestion
              ((error as any)?.message?.toString() ??
              (error as any)?.toString() ??
```

File: packages/extension/src/background/wallet.ts [hunk 2]
```diff
@@ -46,6 +43,10 @@ import {
 } from "./keys/keyDerivation"
 import backupSchema from "./schema/backup.schema"
 
+console.log(hash)
```
Comment (delaaxe) [inline packages/extension/src/background/wallet.ts:46]: Forgot to remove?

File: packages/extension/src/shared/messages/TransactionMessage.ts [hunk 3]
```diff
@@ -55,3 +60,12 @@ export type TransactionMessage =
...
       data: any
     }
+  | {
+      type: "ESTIMATE_DEPLOY_CONTRACT_FEE"
+      data: UniversalDeployerContractPayload
+    }
+  | { type: "ESTIMATE_DEPLOY_CONTRACT_FEE_REJ"; data: { error: string } }
+  | {
+      type: "ESTIMATE_DEPLOY_CONTRACT_FEE_RES"
+      data: any
```
Comment (delaaxe) [inline packages/extension/src/shared/messages/TransactionMessage.ts:70]: Can we type this correctly here?

File: packages/extension/src/ui/features/actions/feeEstimation/DeployContractFeeEstimation.tsx [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+  const amountCurrencyValue = useTokenAmountToCurrencyValue(
+    feeToken,
+    fee?.accountDeploymentFee,
+  )
+  const suggestedMaxFeeCurrencyValue = useTokenAmountToCurrencyValue(
+    feeToken,
+    fee?.maxADFee,
+  )
+
+  return (
```
Comment (delaaxe) [inline packages/extension/src/ui/features/actions/feeEstimation/DeployContractFeeEstimation.tsx:96]: Should be spllit into smaller components

File: packages/extension/src/ui/features/settings/DeveloperSettings/ClassHashInputActions.tsx [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+                            mt={1}
+                            _groupHover={{ color: "white" }}
+                          >
+                            {formatTruncatedAddress(
+                              transaction.meta?.subTitle || "",
+                            )}
+                          </P4>
+                        </Flex>
+                      </Flex>
+                    ),
```
Comment (delaaxe) [inline packages/extension/src/ui/features/settings/DeveloperSettings/ClassHashInputActions.tsx:140]: could be own component

File: packages/extension/src/ui/features/settings/DeveloperSettings/DeploySmartContractForm.tsx [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+      )
+      const constructorAbi = abi?.find((item) => item.type === "constructor")
+      setParameterFields(
+        constructorAbi?.inputs.map((input: AbiEntry) => ({
+          name: input.name,
+          type: input.type,
+          value: "",
+        })) || [],
+      )
+    } catch (error) {
```
Comment (delaaxe) [inline packages/extension/src/ui/features/settings/DeveloperSettings/DeploySmartContractForm.tsx:134]:
```suggestion
    } catch {
```

File: packages/extension/src/background/transactions/onupdate/declareContract.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import { declaredTransactionsStore } from "../../../shared/udc/store"
+import { UdcTransactionType } from "../../udcAction"
+import { TransactionUpdateListener } from "./type"
+
+export const handleDeclareContractTransaction: TransactionUpdateListener =
+  async (transactions) => {
+    for (const transaction of transactions) {
+      if (
+        transaction.meta?.type === UdcTransactionType.DECLARE_CONTRACT &&
```
Comment (delaaxe) [inline packages/extension/src/background/transactions/onupdate/declareContract.ts:9]: same condition checked twice

File: packages/extension/src/ui/features/actions/feeEstimation/DeployContractFeeEstimation.tsx (hunk 4 above)
Comment (bluecco) [reply packages/extension/src/ui/features/actions/feeEstimation/DeployContractFeeEstimation.tsx:96]: for now I've used `AccountDeploymentFeeEstimation.tsx` as base.

It should be definitely refactored as you suggest.
@delaaxe @janek26 I can do this in the current pr or in a followup

File: packages/extension/src/background/wallet.ts (hunk 2 above)
Comment (bluecco) [reply packages/extension/src/background/wallet.ts:46]: yep 😓

**Comment by janek26** [comment]
haha, thanks for the review @delaaxe ! 😂

File: packages/extension/src/ui/features/actions/feeEstimation/DeployContractFeeEstimation.tsx (hunk 4 above)
Comment (janek26) [reply packages/extension/src/ui/features/actions/feeEstimation/DeployContractFeeEstimation.tsx:96]: follow up, pls create a jira ticket to unify all Transaction/Declare/Deploy/etc approval screens

File: packages/dapp/src/components/TokenDapp.tsx [hunk 8]
```diff
@@ -1,7 +1,7 @@
 import { SessionAccount, createSession } from "@argent/x-sessions"
 import { FC, useEffect, useState } from "react"
 import { Abi, AccountInterface, Contract } from "starknet"
-import { genKeyPair, getStarkKey } from "starknet/dist/utils/ellipticCurve"
+import { ec } from "starknet"
```
Comment (janek26) [inline packages/dapp/src/components/TokenDapp.tsx:4]: dup import

File: packages/extension/src/shared/call/erc721TransferCall.ts [hunk 9]
```diff
@@ -1,15 +1,16 @@
 import { Call, validateAndParseAddress } from "starknet"
-import { BigNumberish } from "starknet/dist/utils/number"
-import { Uint256, uint256ToBN } from "starknet/dist/utils/uint256"
+import { number, uint256 } from "starknet"
```
Comment (janek26) [inline packages/extension/src/shared/call/erc721TransferCall.ts:2]: same

File: packages/extension/src/shared/messages/UdcMessage.ts [hunk 10]
```diff
@@ -31,4 +31,23 @@ export type UdpMessage =
...
+      type: "DEPLOY_CONTRACT_ACTION_FAILED"
+      data: { actionHash: string; error?: string }
+    }
+  | {
+      type: "FETCH_CONSTRUCTOR_PARAMS"
+      data: { classHash: string; networkId: string }
+    }
+  | {
+      type: "FETCH_CONSTRUCTOR_PARAMS_RES"
+      data: { contract: any } /* TODO:type */
```
Comment (janek26) [inline packages/extension/src/shared/messages/UdcMessage.ts:48]: please type before merge

File: packages/extension/src/ui/features/actions/ActionScreen.tsx [hunk 11]
```diff
@@ -283,6 +284,50 @@ export const ActionScreen: FC = () => {
...
         />
       )
+    case "DEPLOY_CONTRACT_ACTION":
+      return (
+        <>
+          <ApproveDeployContractScreen
+            actionHash={action.meta.hash}
+            deployPayload={action.payload}
+            onSubmit={async () => {
+              analytics.track("signedTransaction", {
```
Comment (janek26) [inline packages/extension/src/ui/features/actions/ActionScreen.tsx:294]: we should probably introduce new tracking events for this

File: packages/extension/src/ui/features/actions/ApproveDeployContractScreen.tsx [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+const ApproveDeployContractScreen: FC<ApproveDeployContractScreenProps> = ({
+  selectedAccount,
+  actionHash,
+  deployPayload: { classHash, salt, unique, constructorCalldata },
+  onSubmit,
+  ...props
+}) => {
+  usePageTracking("signTransaction", {
```
Comment (janek26) [inline packages/extension/src/ui/features/actions/ApproveDeployContractScreen.tsx:31]: same

File: packages/extension/src/ui/services/backgroundTransactions.ts [hunk 13]
```diff
@@ -58,6 +58,24 @@ export const getDeclareContractEstimatedFee = async (data: DeclareContract) => {
...
+) => {
+  sendMessage({ type: "ESTIMATE_DEPLOY_CONTRACT_FEE", data })
+
+  const response = await Promise.race([
+    waitForMessage("ESTIMATE_DEPLOY_CONTRACT_FEE_RES"),
+    waitForMessage("ESTIMATE_DEPLOY_CONTRACT_FEE_REJ"),
+  ])
+
+  if ("error" in response) {
+    console.error(response.error)
```
Comment (janek26) [inline packages/extension/src/ui/services/backgroundTransactions.ts:72]: you want to keep this?

File: packages/extension/src/shared/messages/UdcMessage.ts (hunk 10 above)
Comment (bluecco) [reply packages/extension/src/shared/messages/UdcMessage.ts:48]: yep, missed that

File: packages/extension/src/ui/services/backgroundTransactions.ts (hunk 13 above)
Comment (bluecco) [reply packages/extension/src/ui/services/backgroundTransactions.ts:72]: no, forgot to remove

File: packages/extension/src/ui/features/actions/ActionScreen.tsx (hunk 11 above)
Comment (bluecco) [reply packages/extension/src/ui/features/actions/ActionScreen.tsx:294]: `declareTransaction` would be good?

File: packages/extension/src/ui/features/actions/ApproveDeployContractScreen.tsx (hunk 12 above)
Comment (bluecco) [reply packages/extension/src/ui/features/actions/ApproveDeployContractScreen.tsx:31]: `deployTransaction` would be good?

Comment (bluecco) [reply packages/extension/src/ui/features/actions/ApproveDeployContractScreen.tsx:31]: same as comment before, used the same as transaction and accountDeploy

File: packages/extension/src/ui/features/actions/feeEstimation/DeployContractFeeEstimation.tsx (hunk 4 above)
Comment (bluecco) [reply packages/extension/src/ui/features/actions/feeEstimation/DeployContractFeeEstimation.tsx:96]: ticket created, will work on it just after the release of this @janek26

File: packages/extension/src/ui/features/actions/ApproveDeployContractScreen.tsx (hunk 12 above)
Comment (bluecco) [reply packages/extension/src/ui/features/actions/ApproveDeployContractScreen.tsx]: actually `signDeployTransaction` is better (so there is also the context)
same for above

**Comment by bluecco** [comment]
@janek26 changes to comments pushed

**Comment by janek26** [comment]
conflicts now :)

**Review by janek26 — APPROVED** [review]

**Comment by bluecco** [comment]
@janek26 solved

**Pull request merged**